### PR TITLE
Thyra,Belos,Anasazi: Fix valgrind findings

### PR DESCRIPTION
**Pull request by cgcgcg**
@trilinos/thyra @trilinos/belos @trilinos/anasazi 

## Motivation
Minor changes to Thyra, Belos, Anasazi unit tests that were flagged by Valgrind:
https://sems-cdash-son.sandia.gov/cdash/viewDynamicAnalysis.php?buildid=88114
https://sems-cdash-son.sandia.gov/cdash/viewDynamicAnalysis.php?buildid=88098
https://sems-cdash-son.sandia.gov/cdash/viewDynamicAnalysis.php?buildid=88099

### CODE REVIEW

File: packages/thyra/core/test/operator_vector/DefaultProductVectorSpace_UnitTests.cpp [hunk 1]
```diff
@@ -471,6 +474,7 @@ TEUCHOS_UNIT_TEST_TEMPLATE_1_DECL( DefaultProductVector,
     createSpmdVectorSpace<Scalar>(g_localDim);
 
   const RCP<const VectorBase<Scalar> > v = createMember(vs);
+  Teuchos::rcp_const_cast<VectorBase<Scalar> >(v)->assign(0.);
```
Comment (bartlettroscoe) [inline packages/thyra/core/test/operator_vector/DefaultProductVectorSpace_UnitTests.cpp:477]: Same as above

```suggestion
  const RCP<VectorBase<Scalar> > v = createMember(vs);
  Thyra::put_scalar<Scalar>(0.0, v.ptr());
```

**Review by jhux2 — DISMISSED** [review]
lgtm

File: packages/thyra/core/test/operator_vector/DefaultSpmdMultiVector_UnitTests.cpp [hunk 2]
```diff
@@ -101,7 +101,7 @@ TEUCHOS_UNIT_TEST_TEMPLATE_1_DECL( DefaultSpmdMultiVector, defaultConstruct,
   RCP<const MultiVectorBase<Scalar> > mv = createMembers(*vs, g_numCols);
   Teuchos::Array<Scalar> mv_sums(g_numCols);
   Thyra::sums<Scalar>(*mv, mv_sums());
-  out << "sums(mv) = " << mv_sums << "\n";
+  // out << "sums(mv) = " << mv_sums << "\n";
```
Comment (bartlettroscoe) [inline packages/thyra/core/test/operator_vector/DefaultSpmdMultiVector_UnitTests.cpp:104]: This is just a badly written test.  You should not even read the elements unless they are initialized.  The test should be something like:

```
{
  const RCP<const VectorSpaceBase<Scalar> > vs = createSpmdVectorSpace<Scalar>(g_localDim);
  const RCP<MultiVectorBase<Scalar> > mv = createMembers(*vs, g_numCols);
  Teuchos::Array<Scalar> mv_sums(g_numCols);

#ifndef THYRA_INITIALIZE_VECS_MULTIVECS_WITH_NANS
  Thyra::assign<Scalar>(mv.ptr(), 1.0);   // Can't touch uninitialized memory!
#endif

  Thyra::sums<Scalar>(*mv, mv_sums());
  out << "sums(mv) = " << mv_sums << "\n";

#ifdef THYRA_INITIALIZE_VECS_MULTIVECS_WITH_NANS
  typedef Teuchos::ScalarTraits<Scalar> ST;
  for (int i = 0; i < g_numCols; ++i) {
    TEST_ASSERT(ST::isnaninf(mv_sums[i]));
  }
#endif
}
```

**Review by bartlettroscoe — CHANGES_REQUESTED** [review]
@cgcgcg, thanks for catching these.  Just a few tweaks to make make these tests better (and avoid casting away const (see [ES.50: Don’t cast away `const`](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-casts-const)).

I can make the required changes locally and push to this branch if you would like.  (This was code that I wrote and mistakes that I made).

How is it that you are running valgrind on this code?  Is Trilinos setting up an automated valgrind build?  (Valgrand has not been run on Trilinos in many years and perhaps never on all of Trilinos?)

**Comment by cgcgcg** [comment]
Thanks @bartlettroscoe ! I have applied your changes. The MueLu team is running some nightly builds that are posting here: https://sems-cdash-son.sandia.gov/cdash/index.php?project=Trilinos#!#DynamicAnalysis

**Review by hkthorn — DISMISSED** [review]
LGTM

File: packages/thyra/core/test/operator_vector/DefaultProductVectorSpace_UnitTests.cpp [hunk 3]
```diff
@@ -444,10 +446,11 @@ TEUCHOS_UNIT_TEST_TEMPLATE_1_DECL( DefaultProductVector,
   const RCP<const VectorSpaceBase<Scalar> > vs =
     createProductVectorSpace<Scalar>(g_localDim, numBlocks);
 
-  const RCP<const VectorBase<Scalar> > v = createMember(vs);
+  const RCP<VectorBase<Scalar> > v = createMember(vs);
+  Thyra::put_scalar<Scalar>(0.0, v.ptr());
 
   const RCP<const ProductVectorBase<Scalar> > prod_v =
-    castOrCreateProductVectorBase(v);
+    castOrCreateProductVectorBase(Teuchos::rcp_const_cast<const VectorBase<Scalar> >(v));
```
Comment (bartlettroscoe) [inline packages/thyra/core/test/operator_vector/DefaultProductVectorSpace_UnitTests.cpp:453]: FYI, as mentioned in section "5.7.3 Implicit type conversion problems and shortcomings" and subsection "Implicit conversions failing due to templated function" in [Teuchos C++ Memory Management Classes, Idioms, and Related Topics: The Complete Reference (A Comprehensive Strategy for Safe and Efficient Memory Management in C++ for High Performance Computing)](https://bartlettroscoe.github.io/publications/TeuchosMemoryManagementSAND.pdf), a cleaner way to allow this implicit conversion is to just add the template argument as `castOrCreateProductVectorBase<Scalar>()`.
```suggestion
    castOrCreateProductVectorBase<Scalar>(v);
```

File: packages/thyra/core/test/operator_vector/DefaultProductVectorSpace_UnitTests.cpp [hunk 4]
```diff
@@ -470,10 +473,11 @@ TEUCHOS_UNIT_TEST_TEMPLATE_1_DECL( DefaultProductVector,
   const RCP<const VectorSpaceBase<Scalar> > vs =
     createSpmdVectorSpace<Scalar>(g_localDim);
 
-  const RCP<const VectorBase<Scalar> > v = createMember(vs);
+  const RCP<VectorBase<Scalar> > v = createMember(vs);
+  Thyra::put_scalar<Scalar>(0.0, v.ptr());
 
   const RCP<const ProductVectorBase<Scalar> > prod_v =
-    castOrCreateProductVectorBase(v);
+    castOrCreateProductVectorBase(Teuchos::rcp_const_cast<const VectorBase<Scalar> >(v));
```
Comment (bartlettroscoe) [inline packages/thyra/core/test/operator_vector/DefaultProductVectorSpace_UnitTests.cpp:480]:
```suggestion
    castOrCreateProductVectorBase<Scalar>(v);
```

**Review by bartlettroscoe — APPROVED** [review]
Just a minor suggestion for cleaner code that is a better example of recommended practice for dealing with a common smart pointer deficiency of the C++ language.

**Pull request merged**